File: SOTALog/SOTALog/ViewModels/QRZSyncViewModel.swift
import Foundation
import Observation

@MainActor @Observable
final class QRZSyncViewModel {
    private let database: AppDatabase
    private let qsoRepo: QSORepository
    private let logRepo: LogRepository
    private let historyRepo: CallsignHistoryRepository
    private let refRepo: ReferenceRepository

    var hasAPIKey = false
    var hasCredentials = false
    var unsyncedCount = 0
    var syncStatus: SyncStatus = .idle
    var lastSyncDate: Date?
    var adifExport: String = ""

    enum SyncStatus: Equatable {
        case idle
        case synced
        case uploading(Int, Int)        // (done, total)
        case preparingReferences        // auto-fetching POTA/SOTA ref data
        case downloading(Int)           // QSOs fetched so far
        case importing                  // writing to DB
        case error(String)
    }

    var isBusy: Bool {
        switch syncStatus {
        case .idle, .synced, .error: return false
        default: return true
        }
    }

    var isAllSynced: Bool {
        if case .synced = syncStatus { return true }
        if case .idle = syncStatus { return unsyncedCount == 0 }
        return false
    }

    // Credential testing
    var isTestingCredentials = false
    var apiKeyTestResult: CredentialTestResult?
    var xmlLoginTestResult: CredentialTestResult?

    enum CredentialTestResult {
        case success
        case failure(String)
    }

    init(database: AppDatabase) {
        self.database = database
        self.qsoRepo = QSORepository(database: database)
        self.logRepo = LogRepository(database: database)
        self.historyRepo = CallsignHistoryRepository(database: database)
        self.refRepo = ReferenceRepository(database: database)
    }

    func loadState() async {
        hasAPIKey = KeychainService.load(key: .qrzAPIKey) != nil
        hasCredentials = KeychainService.load(key: .qrzUsername) != nil
        unsyncedCount = (try? await qsoRepo.fetchUnsynced().count) ?? 0
        lastSyncDate = try? await qsoRepo.lastSyncDate()
        if unsyncedCount == 0, lastSyncDate != nil {
            syncStatus = .synced
        }
        await refreshADIF()
    }

    private func refreshADIF() async {
        let logs = (try? await logRepo.fetchAll()) ?? []
        let allQSOs = (try? await qsoRepo.fetchAll()) ?? []
        let qsosByLogId = Dictionary(grouping: allQSOs, by: { $0.logId })
        let sections = logs.compactMap { log -> (Log, [QSO])? in
            guard let id = log.id else { return nil }
            let qsos = qsosByLogId[id] ?? []
            return qsos.isEmpty ? nil : (log, qsos)
        }
        adifExport = ADIFFormatter.encodeFile(sections: sections)
    }

    func saveCredentials(apiKey: String, username: String, password: String) async {
        if !apiKey.isEmpty {
            try? KeychainService.save(key: .qrzAPIKey, value: apiKey)
            hasAPIKey = true
        }
        if !username.isEmpty {
            try? KeychainService.save(key: .qrzUsername, value: username)
            hasCredentials = true
        }
        if !password.isEmpty {
            try? KeychainService.save(key: .qrzPassword, value: password)
        }

        isTestingCredentials = true
        apiKeyTestResult = nil
        xmlLoginTestResult = nil

        async let apiTest: Void = testAPIKey(apiKey)
        async let xmlTest: Void = testXMLLogin(username, password)
        _ = await (apiTest, xmlTest)

        isTestingCredentials = false
    }

    private func testAPIKey(_ apiKey: String) async {
        guard !apiKey.isEmpty else { return }
        do {
            try await QRZLogbookService.testAPIKey(apiKey: apiKey)
            apiKeyTestResult = .success
        } catch {
            apiKeyTestResult = .failure(error.localizedDescription)
        }
    }

    private func testXMLLogin(_ username: String, _ password: String) async {
        guard !username.isEmpty, !password.isEmpty else { return }
        do {
            let sessionKey = try await QRZXMLService.login(username: username, password: password)
            try? KeychainService.save(key: .qrzSessionKey, value: sessionKey)
            xmlLoginTestResult = .success
        } catch {
            xmlLoginTestResult = .failure(error.localizedDescription)
        }
    }

    var allTestsPassed: Bool {
        let apiOk = apiKeyTestResult.map { if case .success = $0 { return true } else { return false } } ?? true
        let xmlOk = xmlLoginTestResult.map { if case .success = $0 { return true } else { return false } } ?? true
        return apiOk && xmlOk
    }

    // MARK: - Upload

    func uploadAll() async {
        guard let apiKey = KeychainService.load(key: .qrzAPIKey) else { return }

        let unsynced: [QSO]
        do {
            unsynced = try await qsoRepo.fetchUnsynced()
        } catch {
            syncStatus = .error(error.localizedDescription)
            return
        }

        guard !unsynced.isEmpty else { return }

        syncStatus = .uploading(0, unsynced.count)

        do {
            let logs = try await logRepo.fetchAll()
            let logMap = Dictionary(uniqueKeysWithValues: logs.compactMap { log in
                log.id.map { ($0, log) }
            })

            var done = 0
            for qso in unsynced {
                let log = qso.logId.flatMap { logMap[$0] }
                let adif = ADIFFormatter.encode(qso: qso, log: log)

                if let qrzLogId = try await QRZLogbookService.uploadQSO(apiKey: apiKey, adifRecord: adif) {
                    if let qsoId = qso.id {
                        try await qsoRepo.markSynced(id: qsoId, qrzLogId: qrzLogId)
                    }
                }

                done += 1
                syncStatus = .uploading(done, unsynced.count)
            }

            unsyncedCount = 0
            syncStatus = .synced
            await refreshADIF()
        } catch {
            AppLog.sync.error("Upload failed: \(error)")
            syncStatus = .error(error.localizedDescription)
        }
    }

    // MARK: - Refresh from QRZ

    func refreshFromQRZ() async {
        guard let apiKey = KeychainService.load(key: .qrzAPIKey) else { return }

        // 0. Pre-check: ensure reference data is available
        do {
            try await ensureReferencesLoaded()
        } catch {
            AppLog.sync.error("Reference fetch failed: \(error)")
            syncStatus = .error("Failed to fetch reference data: \(error.localizedDescription)")
            return
        }

        // 1. Download phase: paginate all QSOs from QRZ
        syncStatus = .downloading(0)
        var allRecords: [[String: String]] = []

        do {
            var afterLogId: Int64 = 0
            var previousMaxLogId: Int64 = -1

            for _ in 0..<200 {
                let result = try await QRZLogbookService.downloadQSOs(apiKey: apiKey, afterLogId: afterLogId)
                let records = ADIFFormatter.decode(result.adif)
                allRecords.append(contentsOf: records)

                syncStatus = .downloading(allRecords.count)
                await Task.yield()

                if result.count < 250 { break }

                // Advance cursor using per-record APP_QRZLOG_LOGID
                let pageMaxLogId = records.compactMap { $0["APP_QRZLOG_LOGID"].flatMap(Int64.init) }.max() ?? 0
                if pageMaxLogId == 0 || pageMaxLogId == previousMaxLogId { break }
                previousMaxLogId = pageMaxLogId
                afterLogId = pageMaxLogId + 1
            }
        } catch {
            AppLog.sync.error("Download failed: \(error)")
            syncStatus = .error(error.localizedDescription)
            return
        }

        // 2. Group phase: load validation dictionaries and group records
        let validPotaRefs: [String: String]
        let validSotaCodes: [String: String]
        do {
            validPotaRefs = try await qsoRepo.loadValidPotaRefs()
            validSotaCodes = try await qsoRepo.loadValidSotaCodes()
        } catch {
            syncStatus = .error("Failed to load reference data")
            return
        }

        let fallbackCallsign = KeychainService.load(key: .qrzUsername)?.uppercased()
        let grouped = SyncImporter.groupByActivation(
            records: allRecords,
            fallbackCallsign: fallbackCallsign,
            validPotaRefs: validPotaRefs,
            validSotaCodes: validSotaCodes
        )

        // 3. Import phase: atomic write
        syncStatus = .importing
        do {
            let result = try await qsoRepo.fullRefreshImport(
                groupedQSOs: grouped.activations,
                unattachedQSOs: grouped.unattached
            )
            AppLog.sync.info("Refresh complete: \(result.importedCount) QSOs, \(result.activationsCreated) new activations, \(result.activationsReused) reused")

            // 4. Post-import
            try await historyRepo.rebuildFromQSOTable()
            try await qsoRepo.saveLastSyncedQRZLogId(0)
            lastSyncDate = Date()

            unsyncedCount = (try? await qsoRepo.fetchUnsynced().count) ?? 0
            syncStatus = unsyncedCount == 0 ? .synced : .idle
            await refreshADIF()
        } catch {
            AppLog.sync.error("Import failed: \(error)")
            syncStatus = .error(error.localizedDescription)
        }
    }

    // MARK: - Reference Data

    private func ensureReferencesLoaded() async throws {
        let potaMeta = try await refRepo.fetchMetadata(key: "potaParks")
        let sotaMeta = try await refRepo.fetchMetadata(key: "sotaSummits")

        let needsPota = (potaMeta?.recordCount ?? 0) == 0
        let needsSota = (sotaMeta?.recordCount ?? 0) == 0

        guard needsPota || needsSota else { return }

        syncStatus = .preparingReferences

        if needsPota {
            let parks = try await POTAParkService.fetchAllParks()
            try await refRepo.deleteAllParks()
            try await refRepo.importParks(parks)
            try await refRepo.saveMetadata(ReferenceMetadata(
                key: "potaParks",
                lastRefreshed: Date(),
                recordCount: parks.count
            ))
        }

        if needsSota {
            let summits = try await SOTASummitService.fetchSummits()
            try await refRepo.deleteAllSummits()
            try await refRepo.importSummits(summits)
            try await refRepo.saveMetadata(ReferenceMetadata(
                key: "sotaSummits",
                lastRefreshed: Date(),
                recordCount: summits.count
            ))
        }
    }

    // MARK: - ADIF Export (pre-computed in loadState)
}

File: android/app/src/main/kotlin/com/sotalog/android/ui/sync/QRZSyncViewModel.kt
package com.sotalog.android.ui.sync

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sotalog.android.data.local.database.dao.LogDao
import com.sotalog.android.data.local.database.dao.QSODao
import com.sotalog.android.data.local.database.dao.ReferenceDao
import com.sotalog.android.data.local.preferences.CredentialStore
import com.sotalog.android.data.local.preferences.LocationService
import com.sotalog.android.data.remote.api.POTAParkApi
import com.sotalog.android.data.remote.api.QRZLogbookApi
import com.sotalog.android.data.remote.api.QRZLookupApi
import com.sotalog.android.data.remote.api.SOTASummitApi
import com.sotalog.android.data.repositories.ReferenceRepository
import com.sotalog.android.data.repositories.SummitParkNetworkRepository
import com.sotalog.android.domain.models.Log
import com.sotalog.android.domain.models.POTAPark
import com.sotalog.android.domain.models.QSO
import com.sotalog.android.domain.models.ReferenceMetadata
import com.sotalog.android.domain.models.SOTASummit
import com.sotalog.android.domain.services.ADIFFormatter
import com.sotalog.android.domain.services.SyncImporter
import com.sotalog.android.ui.logging.ADIFFile
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Date
import javax.inject.Inject

sealed class SyncStatus {
    data object Idle : SyncStatus()
    data object Synced : SyncStatus()
    data class Uploading(val done: Int, val total: Int) : SyncStatus()
    data object PreparingReferences : SyncStatus()
    data class Downloading(val count: Int) : SyncStatus()
    data object Importing : SyncStatus()
    data class Error(val message: String) : SyncStatus()
}

sealed class CredentialTestResult {
    data object Success : CredentialTestResult()
    data class Failure(val message: String) : CredentialTestResult()
}

@HiltViewModel
class QRZSyncViewModel @Inject constructor(
    private val logDao: LogDao,
    private val qsoDao: QSODao,
    private val referenceDao: ReferenceDao,
    private val credentialStore: CredentialStore,
    private val qrzLogbookApi: QRZLogbookApi,
    private val qrzLookupApi: QRZLookupApi,
    private val potaParkApi: POTAParkApi,
    private val sotaSummitApi: SOTASummitApi,
    private val summitParkNetworkRepo: SummitParkNetworkRepository,
    private val referenceRepo: ReferenceRepository,
    private val locationService: LocationService,
) : ViewModel() {

    private val _hasAPIKey = MutableStateFlow(false)
    val hasAPIKey: StateFlow<Boolean> = _hasAPIKey.asStateFlow()

    private val _hasCredentials = MutableStateFlow(false)
    val hasCredentials: StateFlow<Boolean> = _hasCredentials.asStateFlow()

    private val _username = MutableStateFlow<String?>(null)
    val username: StateFlow<String?> = _username.asStateFlow()

    private val _unsyncedCount = MutableStateFlow(0)
    val unsyncedCount: StateFlow<Int> = _unsyncedCount.asStateFlow()

    private val _syncStatus = MutableStateFlow<SyncStatus>(SyncStatus.Idle)
    val syncStatus: StateFlow<SyncStatus> = _syncStatus.asStateFlow()

    private val _lastSyncDate = MutableStateFlow<Date?>(null)
    val lastSyncDate: StateFlow<Date?> = _lastSyncDate.asStateFlow()

    private val _adifExport = MutableStateFlow(ADIFFile("", ""))
    val adifExport: StateFlow<ADIFFile> = _adifExport.asStateFlow()

    private val _parkCount = MutableStateFlow(0)
    val parkCount: StateFlow<Int> = _parkCount.asStateFlow()

    private val _summitCount = MutableStateFlow(0)
    val summitCount: StateFlow<Int> = _summitCount.asStateFlow()

    private val _isParkLoading = MutableStateFlow(false)
    val isParkLoading: StateFlow<Boolean> = _isParkLoading.asStateFlow()

    private val _isSummitLoading = MutableStateFlow(false)
    val isSummitLoading: StateFlow<Boolean> = _isSummitLoading.asStateFlow()

    // Credential testing
    private val _isTestingCredentials = MutableStateFlow(false)
    val isTestingCredentials: StateFlow<Boolean> = _isTestingCredentials.asStateFlow()

    private val _apiKeyTestResult = MutableStateFlow<CredentialTestResult?>(null)
    val apiKeyTestResult: StateFlow<CredentialTestResult?> = _apiKeyTestResult.asStateFlow()

    private val _xmlLoginTestResult = MutableStateFlow<CredentialTestResult?>(null)
    val xmlLoginTestResult: StateFlow<CredentialTestResult?> = _xmlLoginTestResult.asStateFlow()

    val isBusy: Boolean
        get() = when (_syncStatus.value) {
            is SyncStatus.Idle, is SyncStatus.Synced, is SyncStatus.Error -> false
            else -> true
        }

    val isAllSynced: Boolean
        get() = when (_syncStatus.value) {
            is SyncStatus.Synced -> true
            is SyncStatus.Idle -> _unsyncedCount.value == 0
            else -> false
        }

    val apiKeyTestPassed: Boolean
        get() = _apiKeyTestResult.value is CredentialTestResult.Success

    val callsignTestPassed: Boolean
        get() = _xmlLoginTestResult.value is CredentialTestResult.Success

    fun loadState() {
        viewModelScope.launch {
            _hasAPIKey.value = credentialStore.load(CredentialStore.QRZ_API_KEY) != null
            _hasCredentials.value = credentialStore.load(CredentialStore.QRZ_USERNAME) != null
            _username.value = credentialStore.load(CredentialStore.QRZ_USERNAME)

            val logs = logDao.getAll()
            var unsynced = 0
            for (log in logs) {
                val logId = log.id ?: continue
                unsynced += qsoDao.getUnsyncedByLogId(logId).size
            }
            _unsyncedCount.value = unsynced

            if (unsynced == 0) {
                _syncStatus.value = SyncStatus.Synced
            }

            refreshADIF()
            loadCounts()
        }
    }

    fun loadCounts() {
        viewModelScope.launch {
            _parkCount.value = referenceDao.getParkCount()
            _summitCount.value = referenceDao.getSummitCount()
        }
    }

    private suspend fun refreshADIF() {
        val logs = logDao.getAll()
        val sections = mutableListOf<Pair<Log, List<QSO>>>()
        val unattachedQSOs = mutableListOf<QSO>()

        for (log in logs) {
            val logId = log.id ?: continue
            val qsos = qsoDao.getByLogId(logId)
            if (qsos.isNotEmpty()) {
                sections.add(log to qsos)
            }
        }

        _adifExport.value = ADIFFile(
            filename = ADIFFormatter.exportAllFilename(),
            content = ADIFFormatter.encodeFile(sections, unattachedQSOs),
        )
    }

    fun saveAPIKey(apiKey: String) {
        if (apiKey.isEmpty()) return
        viewModelScope.launch {
            _isTestingCredentials.value = true
            _apiKeyTestResult.value = null
            testAPIKey(apiKey)
            _isTestingCredentials.value = false

            if (apiKeyTestPassed) {
                credentialStore.save(CredentialStore.QRZ_API_KEY, apiKey)
                _hasAPIKey.value = true
            } else {
                credentialStore.delete(CredentialStore.QRZ_API_KEY)
                _hasAPIKey.value = false
            }
        }
    }

    fun saveCallsignCredentials(username: String, password: String) {
        if (username.isEmpty() || password.isEmpty()) return
        viewModelScope.launch {
            _isTestingCredentials.value = true
            _xmlLoginTestResult.value = null
            testXMLLogin(username, password)
            _isTestingCredentials.value = false

            if (callsignTestPassed) {
                credentialStore.save(CredentialStore.QRZ_USERNAME, username)
                credentialStore.save(CredentialStore.QRZ_PASSWORD, password)
                _hasCredentials.value = true
                _username.value = username
            } else {
                credentialStore.delete(CredentialStore.QRZ_USERNAME)
                credentialStore.delete(CredentialStore.QRZ_PASSWORD)
                _hasCredentials.value = false
                _username.value = null
            }
        }
    }

    private suspend fun testAPIKey(apiKey: String) {
        try {
            val response = qrzLogbookApi.execute(apiKey = apiKey, action = "STATUS")
            if (response.contains("RESULT=OK", ignoreCase = true) || !response.contains("invalid", ignoreCase = true)) {
                _apiKeyTestResult.value = CredentialTestResult.Success
            } else {
                _apiKeyTestResult.value = CredentialTestResult.Failure("Invalid API key")
            }
        } catch (e: Exception) {
            _apiKeyTestResult.value = CredentialTestResult.Failure(e.message ?: "Unknown error")
        }
    }

    private suspend fun testXMLLogin(username: String, password: String) {
        try {
            val response = qrzLookupApi.getSession(username, password)
            val sessionKey = extractXmlTag(response, "Key")
            if (sessionKey != null) {
                credentialStore.save(CredentialStore.QRZ_SESSION_KEY, sessionKey)
                _xmlLoginTestResult.value = CredentialTestResult.Success
            } else {
                val error = extractXmlTag(response, "Error") ?: "Login failed"
                _xmlLoginTestResult.value = CredentialTestResult.Failure(error)
            }
        } catch (e: Exception) {
            _xmlLoginTestResult.value = CredentialTestResult.Failure(e.message ?: "Unknown error")
        }
    }

    fun clearAPIKey() {
        credentialStore.delete(CredentialStore.QRZ_API_KEY)
        _hasAPIKey.value = false
        _syncStatus.value = SyncStatus.Idle
        _apiKeyTestResult.value = null
    }

    fun clearXMLCredentials() {
        credentialStore.delete(CredentialStore.QRZ_USERNAME)
        credentialStore.delete(CredentialStore.QRZ_PASSWORD)
        credentialStore.delete(CredentialStore.QRZ_SESSION_KEY)
        _hasCredentials.value = false
        _username.value = null
        _xmlLoginTestResult.value = null
    }

    fun removeCredentials() {
        clearAPIKey()
        clearXMLCredentials()
    }

    // MARK: - Upload

    fun uploadAll() {
        viewModelScope.launch {
            val apiKey = credentialStore.load(CredentialStore.QRZ_API_KEY) ?: return@launch

            val logs = logDao.getAll()
            val logMap = logs.mapNotNull { log -> log.id?.let { it to log } }.toMap()

            val unsynced = mutableListOf<QSO>()
            for (log in logs) {
                val logId = log.id ?: continue
                unsynced.addAll(qsoDao.getUnsyncedByLogId(logId))
            }

            if (unsynced.isEmpty()) return@launch

            _syncStatus.value = SyncStatus.Uploading(0, unsynced.size)

            try {
                var done = 0
                for (qso in unsynced) {
                    val log = qso.logId?.let { logMap[it] }
                    val adif = ADIFFormatter.encode(qso, log)

                    val response = qrzLogbookApi.execute(
                        apiKey = apiKey,
                        action = "INSERT",
                        adif = adif,
                    )

                    val qrzLogId = extractLogId(response)
                    if (qrzLogId != null && qso.id != null) {
                        qsoDao.update(qso.copy(syncedToQRZ = true, qrzLogId = qrzLogId))
                    }

                    done++
                    _syncStatus.value = SyncStatus.Uploading(done, unsynced.size)
                }

                _unsyncedCount.value = 0
                _syncStatus.value = SyncStatus.Synced
                refreshADIF()
            } catch (e: Exception) {
                _syncStatus.value = SyncStatus.Error(e.message ?: "Upload failed")
            }
        }
    }

    // MARK: - Refresh from QRZ

    fun refreshFromQRZ() {
        viewModelScope.launch {
            val apiKey = credentialStore.load(CredentialStore.QRZ_API_KEY) ?: return@launch

            // 0. Pre-check: ensure reference data is available
            try {
                ensureReferencesLoaded()
            } catch (e: Exception) {
                _syncStatus.value = SyncStatus.Error("Failed to fetch reference data: ${e.message}")
                return@launch
            }

            // 1. Download phase
            _syncStatus.value = SyncStatus.Downloading(0)
            val allRecords = mutableListOf<Map<String, String>>()

            try {
                var afterLogId: Long = 0
                var previousMaxLogId: Long = -1

                for (page in 0 until 200) {
                    val response = qrzLogbookApi.execute(
                        apiKey = apiKey,
                        action = "FETCH",
                        option = "BETWEEN:$afterLogId:999999999999",
                    )

                    val adifPortion = extractAdifFromResponse(response)
                    val records = ADIFFormatter.decode(adifPortion)
                    allRecords.addAll(records)

                    _syncStatus.value = SyncStatus.Downloading(allRecords.size)

                    if (records.size < 250) break

                    val pageMaxLogId = records.mapNotNull { it["APP_QRZLOG_LOGID"]?.toLongOrNull() }.maxOrNull() ?: 0
                    if (pageMaxLogId == 0L || pageMaxLogId == previousMaxLogId) break
                    previousMaxLogId = pageMaxLogId
                    afterLogId = pageMaxLogId + 1
                }
            } catch (e: Exception) {
                _syncStatus.value = SyncStatus.Error(e.message ?: "Download failed")
                return@launch
            }

            // 2. Group phase
            val validPotaRefs = loadValidPotaRefs()
            val validSotaCodes = loadValidSotaCodes()

            val fallbackCallsign = credentialStore.load(CredentialStore.QRZ_USERNAME)?.uppercase()
            val grouped = SyncImporter.groupByActivation(
                records = allRecords,
                fallbackCallsign = fallbackCallsign,
                validPotaRefs = validPotaRefs,
                validSotaCodes = validSotaCodes,
            )

            // 3. Import phase
            _syncStatus.value = SyncStatus.Importing
            try {
                fullRefreshImport(grouped)
                // The worked count is derived from the qso table, so a full refresh
                // needs no count rebuild.
                _lastSyncDate.value = Date()

                var unsynced = 0
                for (log in logDao.getAll()) {
                    val logId = log.id ?: continue
                    unsynced += qsoDao.getUnsyncedByLogId(logId).size
                }
                _unsyncedCount.value = unsynced
                _syncStatus.value = if (unsynced == 0) SyncStatus.Synced else SyncStatus.Idle
                refreshADIF()
            } catch (e: Exception) {
                _syncStatus.value = SyncStatus.Error(e.message ?: "Import failed")
            }
        }
    }

    private suspend fun fullRefreshImport(grouped: SyncImporter.GroupingResult) {
        // Create or reuse logs for each activation
        for ((key, records) in grouped.activations) {
            val existingLogs = logDao.getAll().filter { log ->
                log.date == key.date &&
                    log.potaReference == key.potaReference &&
                    log.sotaReference == key.sotaReference
            }

            val logId: Long
            if (existingLogs.isNotEmpty()) {
                logId = existingLogs.first().id!!
                // Delete existing QSOs for this log (full refresh)
                qsoDao.deleteByLogId(logId)
            } else {
                logId = logDao.insert(
                    Log(
                        createdAt = Date(),
                        date = key.date,
                        myCallsign = key.stationCallsign,
                        myGrid = key.myGrid,
                        potaReference = key.potaReference,
                        sotaReference = key.sotaReference,
                    ),
                )
            }

            for (record in records) {
                val qso = record.qso.copy(
                    id = null,
                    logId = logId,
                    syncedToQRZ = true,
                    qrzLogId = record.rawFields["APP_QRZLOG_LOGID"]?.toLongOrNull(),
                )
                try {
                    qsoDao.insert(qso)
                } catch (_: Exception) {
                    // Skip duplicates
                }
            }
        }

        // Import unattached QSOs
        for (record in grouped.unattached) {
            val qso = record.qso.copy(
                id = null,
                logId = null,
                syncedToQRZ = true,
                qrzLogId = record.rawFields["APP_QRZLOG_LOGID"]?.toLongOrNull(),
            )
            try {
                qsoDao.insert(qso)
            } catch (_: Exception) {
                // Skip duplicates
            }
        }
    }


    private suspend fun loadValidPotaRefs(): Map<String, String> {
        val parks = referenceDao.searchParksByNormalizedPrefix("", 100_000)
        return parks.mapNotNull { park ->
            park.referenceNormalized?.let { it to park.reference }
        }.toMap()
    }

    private suspend fun loadValidSotaCodes(): Map<String, String> {
        val summits = referenceDao.searchSummitsByNormalizedPrefix("", 100_000)
        return summits.mapNotNull { summit ->
            summit.codeNormalized?.let { it to summit.code }
        }.toMap()
    }

    // MARK: - Reference Data

    private suspend fun ensureReferencesLoaded() {
        val potaMeta = referenceDao.getMetadata("potaParks")
        val sotaMeta = referenceDao.getMetadata("sotaSummits")

        val needsPota = (potaMeta?.recordCount ?: 0) == 0
        val needsSota = (sotaMeta?.recordCount ?: 0) == 0

        if (!needsPota && !needsSota) return

        _syncStatus.value = SyncStatus.PreparingReferences

        if (needsPota) {
            downloadParks()
        }

        if (needsSota) {
            downloadSummits()
        }
    }

    fun downloadParks() {
        viewModelScope.launch {
            _isParkLoading.value = true
            _syncStatus.value = SyncStatus.PreparingReferences
            try {
                val parks = summitParkNetworkRepo.fetchPOTAParks()
                referenceDao.deleteAllParks()
                referenceRepo.importParks(parks)

                // Enrich with coordinates from POTA location API
                try {
                    summitParkNetworkRepo.enrichParks(
                        refRepo = referenceRepo,
                        userLatitude = locationService.currentLatitude.value,
                        userLongitude = locationService.currentLongitude.value,
                    )
                } catch (_: Exception) {
                    // Partial enrichment is fine — keep whatever parks were imported
                }

                referenceDao.upsertMetadata(
                    ReferenceMetadata(
                        key = "potaParks",
                        lastRefreshed = Date(),
                        recordCount = referenceDao.getParkCount(),
                    ),
                )
                _parkCount.value = referenceDao.getParkCount()
                _syncStatus.value = SyncStatus.Idle
            } catch (e: Exception) {
                _syncStatus.value = SyncStatus.Error("Failed to download parks: ${e.message}")
            } finally {
                _isParkLoading.value = false
            }
        }
    }

    fun downloadSummits() {
        viewModelScope.launch {
            _isSummitLoading.value = true
            _syncStatus.value = SyncStatus.PreparingReferences
            try {
                val summits = summitParkNetworkRepo.fetchSOTASummits()
                referenceDao.deleteAllSummits()
                referenceRepo.importSummits(summits)
                referenceDao.upsertMetadata(
                    ReferenceMetadata(
                        key = "sotaSummits",
                        lastRefreshed = Date(),
                        recordCount = referenceDao.getSummitCount(),
                    ),
                )
                _summitCount.value = referenceDao.getSummitCount()
                _syncStatus.value = SyncStatus.Idle
            } catch (e: Exception) {
                _syncStatus.value = SyncStatus.Error("Failed to download summits: ${e.message}")
            } finally {
                _isSummitLoading.value = false
            }
        }
    }

    fun exportADIF(): ADIFFile = _adifExport.value

    // MARK: - XML Helpers

    private fun extractXmlTag(xml: String, tag: String): String? {
        val pattern = "<$tag>(.*?)</$tag>"
        val match = Regex(pattern, RegexOption.DOT_MATCHES_ALL).find(xml) ?: return null
        return match.groupValues[1].trim().ifEmpty { null }
    }

    private fun extractLogId(response: String): Long? {
        // QRZ logbook response contains LOGID=<number>
        val match = Regex("LOGID=(\\d+)").find(response)
        return match?.groupValues?.get(1)?.toLongOrNull()
    }

    private fun extractAdifFromResponse(response: String): String {
        // QRZ logbook FETCH response contains ADIF data
        val adifStart = response.indexOf("ADIF=")
        if (adifStart >= 0) {
            return response.substring(adifStart + 5)
        }
        return response
    }
}
